lib/application/widgets/theme/custom_button.dart

```diff
@@ -0,0 +1,276 @@
+
+  final String text;
+
+  /// {@template _CustomElevatedButton.backgroundColor}
+  /// Button's background color swatch.
+  /// {@endtemplate}
+  final MaterialColor? backgroundColor;
+  final VoidCallback? onPressed;
+
+  /// {@template _CustomElevatedButton.leadingAsset}
+  /// Optional leading asset.
+  ///
+  /// Its color will be overriden by the button `TextStyle`.
+  /// {@endtemplate}
+  final String? leadingAsset;
+
+  /// Builds the button background color based on the button state.
+  final Color Function(_ButtonState state) backgroundColorBuilder;
+
+  @override
+  Widget build(BuildContext context) {
+    final textTheme = Theme.of(context).textTheme.button!;
+
+    Widget leadingAssetBuilder(_ButtonState state) => Image.asset(leadingAsset!, color: textTheme.color);
+
+    return _CustomButton(
+      text: text,
+      onPressed: onPressed,
+      isPressedOverlayEnabled: true,
+      backgroundColorBuilder: backgroundColorBuilder,
+      textStyleBuilder: (state) => textTheme,
+      leadingWidgetBuilder: leadingAsset != null ? leadingAssetBuilder : null,
+    );
+  }
+}
+
+/// Shared visual style between all custom button's.
+///
+/// Provides a common and customizable button style that must be used by all custom buttons.
+///
+/// See also:
+///
+///   * [PrimaryElevatedButton] - The primary visually opinionated alternative to [ElevatedButton].
+///   * [SecondaryElevatedButton] - The secondary visually opinionated alternative to [ElevatedButton].
+///   * [CustomTextButton] - The visually opinionated alternative to [TextButton].
+class _CustomButton extends HookWidget {
+  const _CustomButton({
+    required this.text,
+    this.onPressed,
+    this.isPressedOverlayEnabled = false,
+    this.leadingWidgetBuilder,
+    this.backgroundColorBuilder,
+    this.textStyleBuilder,
+  });
+
+  /// The button text.
+  ///
+  /// Its style may be customized by implementing [textStyleBuilder].
+  final String text;
+
+  final VoidCallback? onPressed;
+
+  /// If `true` adds a overlay over the button when it's pressed.
+  final bool isPressedOverlayEnabled;
+
+  /// Builds the button leading widget based on the button `state`.
+  final Widget Function(_ButtonState state)? leadingWidgetBuilder;
+
+  /// Builds the button background color based on the button `state`.
+  ///
+  /// If `null` no background color is added.
+  final Color? Function(_ButtonState state)? backgroundColorBuilder;
+
+  /// Builds the button [text] style based on the button `state`.
+  ///
+  /// If `null` the default text theme is used.
+  final TextStyle? Function(_ButtonState state)? textStyleBuilder;
+
+  @override
+  Widget build(BuildContext context) {
+    final initialState = onPressed != null ? _ButtonState.normal : _ButtonState.disabled;
+    final state = useState(initialState);
+
+    useEffect(() {
+      if (state.value != _ButtonState.pressed) {
+        state.value = onPressed != null ? _ButtonState.normal : _ButtonState.disabled;
+      }
+    });
```

Comment:
Where this particular scenario occurs?

Reply:
I think this is a bad design of some sort, not sure though, but it's odd to me that an effect must run every time the build runs

Reply:
That's the only way I found of replacing the old `void didUpdateWidget(covariant CustomButton oldWidget)` functions, since it's now a `HookWidget`. I haven't found further discussions about how to perform this replacement in the Riverpod repository as well, so I'm not sure if there's something obvious that I'm not seeing or if in this scenario we should be using `StatefulWidget` instead. We need to update the buttons interface once the user taps it, and the only hook that I though it would be able to do that is the `useEffect`, since the tap triggers a rebuild.